ots.1
.TH "OTS" 1 "April 29, 2017" "ots version 1.0" "User Commands"

.SH NAME
.PP
ots - OneTimeSecret.com client yeaaaah

.SH SYNOPSIS
.PP
\fBseki\fR [\fImode\fR] [\fIoptions\fR] [\fIarguments\fR]

.SH DESCRIPTION

.PP
ots is an open-source cli-program to work with OneTimeSecret.com (further Service) API.


.PP
Program is in fact a script written in Python 3.


.PP
OneTimeSecret.com provides a way to send some data securely. It stores Secrets. Secret - simply a message with timer that could be seen only one time. Being \fIRecieved\fR, \fIBurned\fR or having timer (\fITTL\fR) expired, Secret's deleted by Service with no way to restore.


.PP
Each secret has two keys associated with it: \fISecret Key\fR and \fIMetadata Key\fR. \fISecret Key\fR is used to recieve a Secret. \fIMetadata Key\fR is used to check Secret's status. Only Secret's creator should have Metadata Key, as it's used to Burn (that is, to erase) a Secret this Metadata associated with.

Identity is simply a pair EMAIL:APIKEY stored in Configuration File. (see below in Credentials section). With no idenity name provided, program tries to use default. If no default credentials found, anonymous account used. It's used for authentification in Service




.PP
ots provides a dozen of modes to work with:

.TP
-S
share a secret. Default mode

.TP
-R
retrieve a secret

.TP
-C
check availability of service. The only mode with no options and arguments

.TP
-B
burn a secret (Not supported by Service yet)

.TP
-M
check metadata

.TP
-G
generate a secret

.TP
-L
get last metadata (Not supported yet)


.PP
Most modes have common options. If option is somewhat ambiguous (like -k), it's explained in Mode's section.
Common options are:

.TP
-i \fIidentity\fR
use \fIidentity\fR (email, apikey) from Credentials File for authentication

.TP
-e \fIemail\fR
use \fIemail\fR for authentication

.TP
-a \fIapikey\fR
use \fIapikey\fR for authentication

.TP
-u
print url instead of key (-S -M -G )

.TP
-m
print metadata key instead of secret key (-M -G)

.TP
-v
verbose. Print as much information as possible
.TP
-q
quiet. Return only exit code

.PP
Warning: -e and -a options \fIoverride\fR credentials provided via -i option, which could lead them to be incorrect. Use wisely.
.sp


.SH
MODES
The following subsections desribe modes usage and their specific options.

.SS
\fBSHARE A SECRET (-S)\fR

.PP
Default mode that lets you to send a Secret and retrieve URL where one can view this Secret


.TP
usage:
\fBots\fR [-S] [\fIoptions\fR] [\fIarguments\fR]

.PP
Options manage both configuration of a secret and secret's message. Arguments are used as in "echo" command: they concatenated with space as a delimeter. Additional message content provided with options is separated with\fI newline \fR between them.

.PP
Options:

.TP
-f \fIfile\fR
append \fI file \fR content to Secret's message.
.TP
-s \fIstring\fR
append string to Secret list
.TP
-t \fItime\fR
Secret's Time-To-Live. \fItime\fR is either a string in a form "1Y2M3d4h5m6s" (where letters are corresponding time units) or simply a number of seconds.
.TP
-r \fIemail\fR
Recipient's \fIemail\fR where Secret's URL will be send to automatically by Service.
.TP
-p \fIpassphrase\fR
\fIpassphrase\fR is used to encrypt a Secret and used as a password to access Secret's content. Hashed via bcrypt, as stated by Service.
.TP
-c
copy result to clipboard (via xsel or xclip). Implies no -v option (i.e. ignores it).
Not supported yet.




.SS
\fBRETRIEVE A SECRET (-R)\fR

.PP
Mode for retrieving a Secret. I.e. it's message.

.TP
usage:
\fBots\fR -R [\fIoptions\fR] [\fIarguments\fR]

.PP
Secret Key can be provided with option or as an argument (or list of them, if key's happened to have spaces somehow)

.PP
Options:

.TP
-k \fIsecret_key\fR
\fIsecret key\fR of a Secret



.SS
\fBCHECK SERVICE -C\fR

.PP
Outputs Online or Offline depending on Service Response.
.PP
-q option forces program to return zero if Service is online and non-zero value otherwise and provide no output.

.TP
usage:
\fBots\fR -C

.PP
No extra options should be provided.




.SS
\fBBURN A SECRET -B\fR

.PP
Burn a Secret before someone read it. Metadata key is required and passed via -k option or as argument(s).

.TP
usage:
\fBots\fR -B [\fIoptions\fR] [\fIarguments\fR]

.PP
Options:

.TP
-k \fImetadata_key\fR
\fImetadata key\fR of a Secret



.SS
\fBCHECK METADATA -M\fR

.PP
Metadata helps to check secret's various information, e.g. a status:

.IP
\(bu new (not checked via Metadata)
.IP
\(bu viewed (checked via Metadata)
.IP
\(bu received
.IP
\(bu burned

.PP
Without -v common option returns only Secret Key associated with Metadata Key provided

.PP
Options:

.TP
-k \fImetadata_key\fR
\fImetadata key\fR of a Secret


.SS
\fBGENERATE A SECRET -G\fR

.PP
"Generate a short, unique secret. This is useful for temporary passwords, one-time pads, salts, etc."

.PP
Metadata and Secret Keys for this Secret are provided.

.TP
-p \fIpassphrase\fR
\fIpassphrase\fR is used to encrypt a Secret and used as a password to access Secret's content. Hashed via bcrypt, as stated by Service.
.TP
-t \fItime\fR
Secret's Time-To-Live. \fItime\fR is either a string in a form "1Y2M3d4h5m6s" (where letters are corresponding time units) or simply a number of seconds.


.SS
\fBRECEIVE LAST METADATA -L\fR
.PP
NOT SUPPORTED YET


.SH
BUGS
.PP
All bugs can be reported to github page of ots project (https://github.com/yaroslavrogov/ots) as issues

.SH
AUTHOR
.PP
Written by Yaroslav Rogov



.SH
EXIT STATUS
.TP
0
everything is OK
.TP
1
program failed
.TP
2
wrong options/usage
.TP
3
service failed or wrong parameters
.TP
4
secret doesn't exist or was received/burned



ots.py
#!/usr/bin/env python3

from json import loads as jsonparse
from json.decoder import JSONDecodeError
import requests
from requests.auth import HTTPBasicAuth

# Configuration loading
from os.path import expanduser
from configparser import ConfigParser

from time import time, ctime

import re

from getopt import getopt
from getopt import GetoptError as goerror
from sys import argv


'''
	TODO:

		1. perms check and error
		2. Metakeys database
		3. clipboard

		MODES:
			main = share
			generate - ?
			retrieve

'''

def usage():
	USAGE = '''
usage: ots [mode] [options] [arguments]

	'''

configdir='~/.config/ots'

status_url = 'https://onetimesecret.com/api/v1/status'
share_url = 'https://onetimesecret.com/api/v1/share'
generate_url = 'https://onetimesecret.com/api/v1/generate'
retrieve_url = 'https://onetimesecret.com/api/v1/secret/{}' # {} = SECRET_KEY
metadata_url = 'https://onetimesecret.com/api/v1/private/{}' # {} = METADATA_KEY
metadata_recent_url = 'https://onetimesecret.com/api/v1/private/recent'
burn_url = 'https://onetimesecret.com/api/v1/private/{}/burn' # {} = METADATA_KEY

secret_url = 'https://onetimesecret.com/secret/{}' # {} = SECRET_KEY
private_url = 'https://onetimesecret.com/private/{}' # {} = METADATA_KEY

timepat = re.compile(
'''
	(?:([1-9][0-9]*)[C])?\s*
	(?:([1-9][0-9]*)[D])?\s*
	(?:([1-9][0-9]*)[Y])?\s*
	(?:([1-9][0-9]*)[M])?\s*
	(?:([1-9][0-9]*)[W])?\s*
	(?:([1-9][0-9]*)[d])?\s*
	(?:([1-9][0-9]*)[h])?\s*
	(?:([1-9][0-9]*)[m])?\s*
	(?:([1-9][0-9]*)[s]?)?\s*
''',
flags=re.VERBOSE)

def print_dummy(a):
	"""
		Does nothing
		Used as a (print=print_dummy) when -q quiet option provided
	"""
	return

def print_request_error(request, process_name):
	print('[ERROR] Request error during', process_name)
	print('  Code:', request.status_code)
	print('  Response:', request.text)


def api_call(url, data, process_name, auth=None, forcereturn=False):
	r = requests.post(url, data=data, auth=auth)
	# print(r.url)
	if r.status_code != 200:
		print('[ERROR] Request error during', process_name)
		print('  Code:', r.status_code)
		print('  Response:', r.text)
		if not forcereturn:
			return None

	try:
		data = jsonparse(r.text)
	except JSONDecodeError:
		data = {}

	return data

def check_availability():
	r = requests.get(status_url);
	if r.status_code==200:
		status = jsonparse(r.text)['status']
		if status == 'nominal':
			print('Online')
			return True
		print('Offline. Status:', status)
	else:
		print('Offline')
	return False;


def share_secret(data, auth=None):
	"""
		TODO: docstring
	"""
	return api_call(share_url, data, 'sharing a Secret', auth=auth)

def retrieve_secret(data, auth=None):
	return api_call(retrieve_url.format(data['secret_key']), data, 'retrieving a Secret', auth=auth)

def retrieve_metadata(data, auth=None):
	return api_call(metadata_url.format(data['metadata_key']), None, 'retrieving Metadata', auth=auth)

def retrieve_latest_metadata(data, auth=None):
	#TODO
	return

def generate_secret(data, auth=None):
	return api_call(generate_url, data, 'generating a Secret', auth=auth)

def burn_secret(data, auth=None):
	return api_call(burn_url.format(data['metadata_key']), None, 'burning a secret', auth=auth)


config = None
# initialization is before __main__ block below

userdir = expanduser('~')
config = ConfigParser()
r = config.read( ('{}/.config/otsrc'.format(userdir), '{}/otsrc'.format(userdir)) ) # for Unix and Windows
if len(r) == 0:
	c = None

def get_identity(name):
	if config == None:
		return None
	if '' == name:
		name = 'DEFAULT'
	if name not in config:
		return None
	data = config[name]
	if 'email' not in data and 'apikey' not in data:
		return None
	return data['email'], data['apikey']



def parsetime(s):

	"""
		Parses string in a form '1C 2D	3Y4M5W 6d7h8m9s' to time units respectively
			(Centuries, Decades, Years, Months and etc.)

		Returns time in seconds

	"""



	a = list(timepat.match(s).groups())
	for i in range(0,len(a)):
		if a[i] == None:
			a[i] = 0;
		a[i]=int(a[i])

	centuries, decades, years, months, weeks, days, hours, minutes, seconds = a;
	days = (centuries * 100 + decades * 10 + years) * 365.25 + months * 30.4375 + weeks * 7 + days
	seconds = ((days * 24 + hours) * 60 + minutes ) * 60 + seconds

	return int(seconds)


def print_metadata_default(meta, print_recieve=True, print_meta=True):

	"""
		Default function for printing received information about metadata

			meta -- dictionary (parsed from json)

		Keyword arguments:

			print_recieve -- if secret was just created so information about receivement's not needed
			print_meta -- print information about metadata itself: key, expiration date (default: True)
	"""

	t = time()


	# Burn call specific problem
	if 'secret_shortkey' in meta:
		meta = meta['state']

	# print(meta)

	if 'secret_key' in meta and meta['secret_key']:
		print( 'Secret key:', meta['secret_key'] )


	if 'state' in meta and print_recieve: # new, viewed (via metadata) or received (burn may be?)
		"""
			4 possible state:
				new - created and metadata wasn't checked
				viewed - metadata was checked at least once
				received - secret was received
				burned - secret was burned
		"""

		if meta['state'] == 'received':
			print( 'Recieved in {}'.format(ctime(meta['received'])) )
		elif meta['state'] == 'burned':
			print( 'Burned in {}'.format(ctime(meta['updated'])))
		else:
			print('Not Recieved')

	else:
		expired = meta['secret_ttl']<=0;
		print( 'Expired' if expired else 'Expiring:', ctime(t+meta['secret_ttl']) )

	if 'passphrase_required' in meta:
		print( 'Passphrase required' if meta['passphrase_required'] else 'No passphrase' )
	if 'recipient' in meta and meta['recipient']:
		print( 'Recipient:', meta['recipient'] )

	if print_meta:
		mexpired = meta['metadata_ttl']<=0;
		print( 'Metadata key:', meta['metadata_key'] )
		print( 'Expired' if mexpired else 'Expiring:', ctime(t+meta['metadata_ttl']) )











if __name__ == '__main__':

	print_metadata = print_metadata_default

	MODES = 'CSGRMLB'

	try:
		opts, args = getopt(argv[1:],
			MODES #Modes
			+ 'vqi:e:a:k:u' # Common options
			+ 's:p:t:f:r:m' # -S share options
			)
	except goerror as err:
		print(err)
		usage()
		exit(2)

	modes=[ k[0] for k in filter(lambda a: a[0][1:] in MODES and not a[0][2:], opts)]

	if len(modes) == 0:
		mode = '-S'
	else:
		mode = modes[-1]


	if mode == '-C':
		if check_availability():
			exit(0)
		else:
			exit(1)


	key = ''
	identity = list(get_identity('DEFAULT'))
	email = ''
	apikey = ''
	verbose = False
	print_url = False

	for o in opts:
		k = o[0]
		arg = o[1]

		if '-e' == k:
			email = arg
			continue

		if '-a' == k:
			apikey = arg
			continue

		if '-i' == k:
			i = get_identity(arg)
			if i == None:
				print('[ERROR] No identity {} found'.format(arg))
				exit(1)
			identity = i

		if '-q' == k:
			print = print_dummy

		if '-v' == k:
			verbose = True

		if '-u' == k:
			print_url = True



	if len(email) != 0:
		identity[0] = email

	if len(apikey) != 0:
		identity[1] = apikey

	if len(identity[0]) == 0 or len(identity[1]) == 0:
		auth = None
	else:
		auth = HTTPBasicAuth(*identity)




	# Sharing a secret
	if mode == '-S':

		data = {
			'secret' : '',
			'passphrase' : '',
			'ttl' : str(7*24*60*60), # default: 1 week
			'recipient' : ''
		}

		quiet=False
		secret=[]
		print_secret=True

		for o in opts:
			k = o[0]
			arg = o[1]

			if '-f' == k:
				try:
					with open(arg) as f:
						secret.append(f.read())
				except IOError as err:
					print(err)
					exit(1)

			if '-s' == k:
				secret.append(arg)
				continue

			if '-t' == k:
				data['ttl'] = str(parsetime(arg))
				continue

			if '-r' == k:
				data['recipient'] = arg;
				continue

			if '-p' == k:
				data['passphrase'] = arg
				continue

			if '-m' == k:
				print_secret = False
				continue


		if len(args) != 0:
			secret.append(' '.join(args))

		if len(secret) == 0:
			print( 'ERROR: No secret provided' )
			exit(1)

		data['secret']	= '\n'.join(secret)

		# print(data)

		meta = share_secret(data, auth)

		if meta == None:
			exit(3)

		if verbose:
			print_metadata(meta)
		elif print_secret:
			if print_url:
				print(secret_url.format(meta['secret_key']))
			else:
				print(meta['secret_key'])
		else:
			if print_url:
				print(private_url.format(meta['metadata_key']))
			else:
				print(meta['metadata_key'])

		exit(0)




	# Checking a metadata
	if mode == '-M':
		data = {	'metadata_key' : '' }

		metadata_key = key

		if len(metadata_key) == 0:
			metadata_key=' '.join(args)

		if len(metadata_key) == 0:
			print('[ERROR] No metadata key was provided')
			exit(2)

		data['metadata_key'] = metadata_key

		meta = retrieve_metadata(data, auth=auth)

		if meta == None:
			exit(3)

		if verbose:
			print_metadata(meta, print_meta=False)
		elif print_secret:
			if 'secret_key' not in meta:
				print('No secret key')
				exit(4)

			if print_url:
				print(secret_url.format(meta['secret_key']))
			else:
				print(meta['secret_key'])
		else:
			if 'metadata_key' not in meta:
				print('No metadata key')
				exit(3)

			if print_url:
				print(metadata_url.format(meta['metadata_key']))
			else:
				print(meta['metadata_key'])

		exit(0)




	# Retrieve a secret
	if mode == '-R':

		data = { 'secret_key' : '' }

		secret_key = key

		if len(secret_key) == 0:
			secret_key = ' '.join(args)

		if len(secret_key) == 0:
			print('[ERROR] No secret key was provided')
			exit(2)

		data['secret_key'] = secret_key

		response = retrieve_secret(data, auth=auth)

		if response == None:
			exit(3)

		if verbose:
			print('Secret Key:', response['secret_key'])
			print('Value:', response['value'])
		else:
			print(response['value'])

		exit(0)




	# Generate a secret
	if mode == '-G':
		# print( 'Not supported yet' )
		# exit(1)

		data = {
			'passphrase' : '',
			'ttl' : str(7*24*60*60), # default: 1 week
			'recipient' : ''
		}

		print_secret = True

		for o in opts:
			k = o[0]
			arg = o[1]

			if '-t' == k:
				data['ttl'] = str(parsetime(arg))
				continue

			if '-r' == k:
				data['recipient'] = arg;
				continue

			if '-p' == k:
				data['passphrase'] = arg
				continue

			if '-m' == k:
				print_secret = False
				continue


		# print(data)

		meta = generate_secret(data, auth)

		if meta == None:
			exit(3)

		if verbose:
			print_metadata(meta)
		elif print_secret:
			if print_url:
				print(secret_url.format(meta['secret_key']))
			else:
				print(meta['secret_key'])
		else:
			if print_url:
				print(private_url.format(meta['metadata_key']))
			else:
				print(meta['metadata_key'])

		exit(0)



	# Retrieve set of metadata for last secrets
	if mode == '-L':
		print( 'Not supported yet' )
		exit(1)


	# Burn a Secret
	if mode == '-B':

		data = {	'metadata_key' : '' }

		metadata_key = key

		if len(metadata_key) == 0:
			metadata_key=' '.join(args)

		if len(metadata_key) == 0:
			print('[ERROR] No metadata key was provided')
			exit(2)

		data['metadata_key'] = metadata_key

		meta = burn_secret(data, auth=auth)

		if meta == None:
			exit(3)

		# print("Burned")
		print_metadata(meta, print_meta=False)

		exit(0)



	usage()
	exit(1)